lib/group-update.js

```diff
@@ -0,0 +1,335 @@
+                            job.verbose(`----------------------------------------`);
+                            if (errorCount == 0) {
+                                job.verbose(`Successfully updated groups for ${successCount} students, with no errors`);
+                                job.succeed();
+                            } else {
+                                job.verbose(`Successfully updated groups for ${successCount} students`);
+                                job.fail(`Error updating ${errorCount} students`);
+                            }
+                        })
+                        .catch((err) => {
+                            if (output != null) {
+                                job.verbose(output);
+                            }
+                            job.fail('Error processing CSV', err);
+                        });
+                });
+            });
+        });
+    },
+
+    // "feedback" gets mapped to {manual: "XXX"} and overrides the contents of "feedback_json"
+    _getFeedbackOrNull(json) {
+        const feedback_string = json.feedback || null;
+        const feedback_json = json.feedback_json || null;
+        let feedback = null;
+        if (feedback_string != null) {
+            feedback = {manual: feedback_string};
+        }
+        if (feedback_json != null) {
+            let feedback_obj = null;
+            try {
+                feedback_obj = JSON.parse(feedback_json);
+            } catch (e) {
+                throw new Error(`Unable to parse "feedback_json" field as JSON: ${e}`);
+            }
+            if (!_.isPlainObject(feedback_obj)) {
+                throw new Error(`Parsed "feedback_json" is not a JSON object: ${feedback_obj}`);
+            }
+            feedback = feedback_obj;
+            if (feedback_string != null) {
+                feedback.manual = feedback_string;
+            }
+        }
+        return feedback;
+    },
+
+    _updateGroupsFromJson(json, assessment_id, authn_user_id, callback) {
+        util.callbackify(async () => {
+            const params = [
+                assessment_id,
+                json.groupName || json.groupname || null,
+                json.uid || json.UID || json.Uid || null,
+            ];
+            await sqldb.callAsync('assessment_groups_update', params);
+        })(callback);
+    },
+
+    /**
+     * Auto generate group settings from input
+     *
+     * @param {number} assessment_id - The assessment to update.
+     * @param {number} user_id - The current user performing the update.
+     * @param {number} authn_user_id - The current authenticated user.
+     * @param {function} callback - A callback(err, job_sequence_id) function.
+     */
+    autoGroups(assessment_id, user_id, authn_user_id, maxnum, minnum, option, callback) {
+        if(maxnum < 2 || minnum < 1 || maxnum < minnum){
+            return callback(new Error('Group Setting Requirements: max > 1; min > 0; max > min'));
```

Comment:
The error message does not exactly match the check. For example, the error message says that we can't have `max == min`, but the `if` statement allows this case.